File: app/src/main/java/org/ole/planet/myplanet/ui/viewer/AudioPlayerActivity.java
package org.ole.planet.myplanet.ui.viewer;

import android.os.Bundle;
import android.view.MenuItem;
import android.view.View;

import androidx.appcompat.app.AppCompatActivity;

import com.example.jean.jcplayer.JcPlayerManagerListener;
import com.example.jean.jcplayer.general.JcStatus;
import com.example.jean.jcplayer.model.JcAudio;

import org.jetbrains.annotations.NotNull;
import org.ole.planet.myplanet.R;
import org.ole.planet.myplanet.databinding.ActivityAudioPlayerBinding;
import org.ole.planet.myplanet.utilities.Utilities;

import java.io.File;
import java.util.ArrayList;
import java.util.regex.Matcher;
import java.util.regex.Pattern;

public class AudioPlayerActivity extends AppCompatActivity implements JcPlayerManagerListener {
    private ActivityAudioPlayerBinding activityAudioPlayerBinding;
    ArrayList<JcAudio> jcAudios;
    boolean isFullPath;
    String filePath;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        activityAudioPlayerBinding = ActivityAudioPlayerBinding.inflate(getLayoutInflater());
        setContentView(activityAudioPlayerBinding.getRoot());
        filePath = getIntent().getStringExtra("TOUCHED_FILE");
        jcAudios = new ArrayList<>();
        isFullPath = getIntent().getBooleanExtra("isFullPath", false);
        if (filePath.matches(".*[0-9a-fA-F]{8}-[0-9a-fA-F]{4}-[0-9a-fA-F]{4}-[0-9a-fA-F]{4}-[0-9a-fA-F]{12}//.*")) {
            playRecordedAudio();
        } else {
            playDownloadedAudio();
        }
    }

    private void playDownloadedAudio() {
        String fullPath;
        if (isFullPath) {
            fullPath = filePath;
        } else {
            File basePath = getExternalFilesDir(null);
            fullPath = new File(basePath, "ole/" + filePath).getAbsolutePath();
        }

        jcAudios.add(JcAudio.createFromFilePath(fullPath));
        initializeJCPlayer();
    }

    private void playRecordedAudio() {
        Pattern uuidPattern = Pattern.compile("^[0-9a-fA-F]{8}-[0-9a-fA-F]{4}-[0-9a-fA-F]{4}-[0-9a-fA-F]{4}-[0-9a-fA-F]{12}/");
        Matcher matcher = uuidPattern.matcher(filePath);

        if (matcher.find()) {
            filePath = filePath.substring(matcher.group().length());
        }
        jcAudios.add(JcAudio.createFromFilePath(filePath));
        initializeJCPlayer();
    }

    private void initializeJCPlayer() {
        activityAudioPlayerBinding.jcplayer.initPlaylist(jcAudios, null);
        View rootView = activityAudioPlayerBinding.jcplayer.getRootView();
        rootView.findViewById(R.id.btnNext).setVisibility(View.GONE);
        rootView.findViewById(R.id.btnPrev).setVisibility(View.GONE);
        rootView.findViewById(R.id.btnRepeatOne).setVisibility(View.GONE);
    }

    @Override
    public boolean onOptionsItemSelected(MenuItem item) {
        if (item.getItemId() == android.R.id.home) {
            finish();
        }
        return super.onOptionsItemSelected(item);
    }

    @Override
    protected void onPause() {
        super.onPause();
        if (activityAudioPlayerBinding.jcplayer != null && activityAudioPlayerBinding.jcplayer.isPlaying()) {
            activityAudioPlayerBinding.jcplayer.pause();
        }
    }

    @Override
    protected void onStop() {
        super.onStop();
        if (activityAudioPlayerBinding.jcplayer != null) activityAudioPlayerBinding.jcplayer.kill();
    }

    @Override
    protected void onResume() {
        super.onResume();
        if (activityAudioPlayerBinding.jcplayer != null && jcAudios.size() > 0) {
            activityAudioPlayerBinding.jcplayer.playAudio(jcAudios.get(0));
        }
    }

    @Override
    public void onCompletedAudio() {
    }

    @Override
    public void onContinueAudio(@NotNull JcStatus jcStatus) {
    }

    @Override
    public void onJcpError(@NotNull Throwable throwable) {
        Utilities.toast(this, getString(R.string.unable_to_play_audio));
    }

    @Override
    public void onPaused(@NotNull JcStatus jcStatus) {
    }

    @Override
    public void onPlaying(@NotNull JcStatus jcStatus) {
    }

    @Override
    public void onPreparedAudio(@NotNull JcStatus jcStatus) {
    }

    @Override
    public void onStopped(@NotNull JcStatus jcStatus) {
    }

    @Override
    public void onTimeChanged(@NotNull JcStatus jcStatus) {
    }
}

File: app/src/main/java/org/ole/planet/myplanet/ui/viewer/AudioPlayerActivity.kt
package org.ole.planet.myplanet.ui.viewer

import android.content.res.Configuration
import android.os.Bundle
import android.view.MenuItem
import android.view.View
import android.widget.FrameLayout
import android.widget.ImageButton
import android.widget.ImageView
import androidx.appcompat.app.AppCompatActivity
import androidx.core.content.ContextCompat
import androidx.media3.common.MediaItem
import androidx.media3.common.PlaybackException
import androidx.media3.common.Player
import androidx.media3.common.util.UnstableApi
import androidx.media3.exoplayer.ExoPlayer
import androidx.media3.ui.DefaultTimeBar
import com.bumptech.glide.Glide
import java.io.File
import java.util.regex.Pattern
import org.ole.planet.myplanet.R
import org.ole.planet.myplanet.databinding.ActivityAudioPlayerBinding
import org.ole.planet.myplanet.utilities.EdgeToEdgeUtils
import org.ole.planet.myplanet.utilities.FileUtils
import org.ole.planet.myplanet.utilities.Utilities

class AudioPlayerActivity : AppCompatActivity() {

    private lateinit var binding: ActivityAudioPlayerBinding
    private var exoPlayer: ExoPlayer? = null
    private var filePath: String? = null
    private var isFullPath = false
    private lateinit var playButton: ImageButton
    private lateinit var pauseButton: ImageButton


    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        binding = ActivityAudioPlayerBinding.inflate(layoutInflater)
        setContentView(binding.root)
        EdgeToEdgeUtils.setupEdgeToEdge(this, binding.root)

        supportActionBar?.setDisplayHomeAsUpEnabled(true)
        filePath = intent.getStringExtra("TOUCHED_FILE")
        isFullPath = intent.getBooleanExtra("isFullPath", false)

        val extractedFileName = FileUtils.nameWithoutExtension(filePath).toString()
        val resourceTitle = intent.getStringExtra("RESOURCE_TITLE") ?: "Unknown Artist"

        supportActionBar?.title = "Audio Player"
        supportActionBar?.subtitle = extractedFileName

        binding.trackTitle.text = extractedFileName
        binding.artistName.text = resourceTitle
        playButton = binding.playerView.findViewById(R.id.exo_play)
        pauseButton = binding.playerView.findViewById(R.id.exo_pause)

        val overlay = binding.playerView.findViewById<FrameLayout>(R.id.exo_overlay)


        val blurredImageView = ImageView(this).apply {
            layoutParams = FrameLayout.LayoutParams(
                FrameLayout.LayoutParams.MATCH_PARENT,
                FrameLayout.LayoutParams.MATCH_PARENT
            )
            scaleType = ImageView.ScaleType.CENTER_CROP
        }

        Glide.with(this)
            .load(getThemeBackground()) // or from URL or filePath
            .into(blurredImageView)

        overlay.addView(blurredImageView, 0)
        val controller = binding.playerView.findViewById<View>(R.id.exo_controller)
        controller?.setBackgroundColor(android.graphics.Color.TRANSPARENT)


        initializeExoPlayer()

        setupPlayPauseButtons()
    }

    @androidx.annotation.OptIn(UnstableApi::class)
    private fun initializeExoPlayer() {
        val fullPath = resolveFullPath(filePath)

        try {
            exoPlayer = ExoPlayer.Builder(this).build().also { player ->
                binding.playerView.player = player
                player.setMediaItem(MediaItem.fromUri(fullPath))
                player.prepare()
                player.playWhenReady = true

                val controller = binding.playerView.findViewById<View>(R.id.exo_controller)
                controller?.setBackgroundColor(android.graphics.Color.TRANSPARENT)

                val timeBar = binding.playerView.findViewById<DefaultTimeBar>(
                    androidx.media3.ui.R.id.exo_progress
                )
                timeBar?.apply {
                    setPlayedColor(ContextCompat.getColor(this@AudioPlayerActivity, R.color.daynight_textColor))
                    setScrubberColor(ContextCompat.getColor(this@AudioPlayerActivity, R.color.daynight_textColor))
                    setBufferedColor(ContextCompat.getColor(this@AudioPlayerActivity, R.color.hint_color))
                    setUnplayedColor(ContextCompat.getColor(this@AudioPlayerActivity, R.color.disable_color))
                }

                player.addListener(object : Player.Listener {
                    override fun onPlayerError(error: PlaybackException) {
                        Utilities.toast(this@AudioPlayerActivity, "Unable to play audio.")
                    }

                    override fun onPlaybackStateChanged(playbackState: Int) {
                        if (playbackState == Player.STATE_ENDED) {
                            player.seekTo(0)
                        }
                    }
                })
            }
        } catch (e: Exception) {
            Utilities.toast(this, "Unable to play audio.")
        }
    }

    private fun resolveFullPath(originalPath: String?): String {
        return if (isFullPath) {
            originalPath ?: ""
        } else {
            val processedPath = originalPath?.let {
                val uuidPattern = Pattern.compile("^[0-9a-fA-F]{8}-[0-9a-fA-F]{4}-[0-9a-fA-F]{4}-[0-9a-fA-F]{4}-[0-9a-fA-F]{12}/")
                val matcher = uuidPattern.matcher(it)
                if (matcher.find()) it.substring(matcher.end()) else it
            }

            File(getExternalFilesDir(null), "ole/$processedPath").absolutePath
        }
    }

    private fun getThemeBackground(): Int {
        val isDarkMode = resources.configuration.uiMode and
                Configuration.UI_MODE_NIGHT_MASK == Configuration.UI_MODE_NIGHT_YES
        return if (isDarkMode) R.drawable.bg_player_dark else R.drawable.bg_player_white
    }

    private fun setupPlayPauseButtons() {
        playButton.setOnClickListener {
            playAudio()
        }

        pauseButton.setOnClickListener {
            pauseAudio()
        }
    }

    private fun playAudio() {
        exoPlayer?.play()
        playButton.visibility = View.GONE
        pauseButton.visibility = View.VISIBLE
    }

    private fun pauseAudio() {
        exoPlayer?.pause()
        pauseButton.visibility = View.GONE
        playButton.visibility = View.VISIBLE
    }



    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        if (item.itemId == android.R.id.home) {
            finish()
            return true
        }
        return super.onOptionsItemSelected(item)
    }


    override fun onPause() {
        super.onPause()
        pauseAudio()
    }

    override fun onDestroy() {
        binding.playerView.player = null
        exoPlayer?.release()
        exoPlayer = null
        super.onDestroy()
    }
}
